### PR TITLE
ROC-4525: CCD data migration implementation

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java [hunk 1]
```diff
@@ -45,33 +62,213 @@ public void migrate() {
         AtomicInteger updatedClaims = new AtomicInteger(0);
         AtomicInteger failedMigrations = new AtomicInteger(0);
 
-        notMigratedClaims.forEach(claim -> {
-            try {
-                logger.info("\t\t start migrating claim: " + claim.getReferenceNumber());
+        notMigratedClaims.sort(Comparator.comparing(Claim::getId));
```
Comment (sabahirfan) [inline ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: We need to process claims in reverse order. `Comparator.comparing(Claim::getId).reversed()`

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java [hunk 2]
```diff
@@ -45,33 +62,213 @@ public void migrate() {
...
+
+        CaseDetails ccdCase = Optional.ofNullable(caseDetails).orElseGet(() ->
+            coreCaseDataService.getCcdIdByReferenceNumber(user, claim.getReferenceNumber()).orElse(null));
+
+        if (ccdCase == null) {
+            return;
+        }
+
+        try {
+            for (CaseEvent event : CaseEvent.values()) {
```
Comment (sabahirfan) [inline ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: Prefer not to rely on the order of enums to decide what event to fire next. I think we should be able to decide based on the current state of claim in CCD.

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java [hunk 3]
```diff
@@ -45,33 +62,213 @@ public void migrate() {
...
+
+                    coreCaseDataService.overwrite(user, ccdCase.getId(), claim, event);
+                    logger.info("\t\t claim exists - overwrite");
+                    updatedClaims.incrementAndGet();
+                }
+            }
+//Enable below line for final run on prod
+// claimRepository.markAsMigrated(claim.getId());
+        } catch (Exception e) {
+            logger.info(e.getMessage(), e);
```
Comment (karthikkannankainos) [inline ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: logger in error when its exception

Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: done

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java (hunk 1 above)
Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: done

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java (hunk 2 above)
Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: can test later with different implementation, but we don't have timestamp for all event, so it is tough to find the sequence

Comment (sabahirfan) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: Ok, we can defer that and will implement if there is any issue found in dry run.

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/idam/models/CaseEvent.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package uk.gov.hmcts.cmc.ccd.migration.idam.models;
+
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/idam/models/CaseEvent.java]: will do in other PR

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java [hunk 5]
```diff
@@ -45,33 +62,213 @@ public void migrate() {
...
+
+    private void updateCase(
+        User user,
+        AtomicInteger updatedClaims,
+        AtomicInteger failedMigrations,
+        Claim claim,
+        CaseDetails caseDetails
+    ) {
+
+        CaseDetails ccdCase = Optional.ofNullable(caseDetails).orElseGet(() ->
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: there can be few cases

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/ccd/services/CoreCaseDataService.java [hunk 6]
```diff
@@ -59,57 +60,59 @@ public CoreCaseDataService(
...
-            migrateCoreCaseDataService.update(user.getAuthorisation(), eventRequestData, ccdId, claim);
+            migrateCoreCaseDataService
+                .update(user.getAuthorisation(), eventRequestData, caseId, claim);
         } catch (Exception exception) {
             throw new OverwriteCaseException(
                 String.format(
                     "Failed updating claim in CCD store for claim %s on event %s",
                     claim.getReferenceNumber(),
-                    EventType.MIGRATED_FROM_CLAIMSTORE_UPDATE), exception
+                    event), exception
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/ccd/services/CoreCaseDataService.java]: doesn't matter, it is for internal logging

[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/ccd/services/CoreCaseDataService.java]: there is different between printing event (object id) and event value. Also we are printing it in the exception. Its not general logging. So it MATTERS

File: ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply ccd-claim-migration/src/main/java/uk/gov/hmcts/cmc/ccd/migration/services/ClaimMigrator.java]: when we are migrating, there wont be. It cant be :-)

**Pull request merged**